File: DesignPatterns/01_GOF23/DP_examples/Go/03_Behavioral/Interpreter/Calculator.go
// 计算器类，用于构建和解析表达式
package Interpreter

import (
	"strconv"
	"strings"
)

func Calculate(expression string) float64 {
	endExpre := createExpressionTree(expression, 0, len(expression)-1)
	return endExpre.Interpret()
}

func findSplit(expression string, start, end int) int {
	tag := -1
	if expression[start] == '(' && expression[end] == ')' {
		start++
		end--
	}
	isInBracket := 0
	moreGrade := 0
	for i := start; i <= end; i++ {
		if expression[i] == '(' {
			isInBracket++
		} else if expression[i] == ')' {
			isInBracket--
		}
		if (expression[i] == '+' || expression[i] == '-') && isInBracket == 0 {
			moreGrade = 1
			tag = i
		} else if (expression[i] == '*' || expression[i] == '/') && moreGrade == 0 && isInBracket == 0 {
			tag = i
		}
	}
	return tag
}

func createExpressionTree(expression string, start, end int) ArithmeticExpr {
	var ptree ArithmeticExpr
	substr := expression[start : end+1]
	if num, err := strconv.ParseFloat(substr, 64); err == nil {
		ptree = NewNumberExpr(num)
	} else {
		tag := findSplit(expression, start, end)
		if tag < 0 {
			// 处理括号内的数字
			s := strings.ReplaceAll(strings.ReplaceAll(substr, "(", ""), ")", "")
			num, _ := strconv.ParseFloat(s, 64)
			ptree = NewNumberExpr(num)
		} else {
			if expression[tag] == '+' {
				ptree = NewAdditionExpr(
					createExpressionTree(expression, start, tag-1),
					createExpressionTree(expression, tag+1, end),
				)
			} else if expression[tag] == '-' {
				ptree = NewSubtractionExpr(
					createExpressionTree(expression, start, tag-1),
					createExpressionTree(expression, tag+1, end),
				)
			} else if expression[tag] == '*' {
				ptree = NewMultiplicationExpr(
					createExpressionTree(expression, start, tag-1),
					createExpressionTree(expression, tag+1, end),
				)
			} else if expression[tag] == '/' {
				ptree = NewDivisionExpr(
					createExpressionTree(expression, start, tag-1),
					createExpressionTree(expression, tag+1, end),
				)
			}
		}
	}
	return ptree
}

File: DesignPatterns/01_GOF23/DP_examples/Lua/03_Behavioral/Interpreter/Calculator.lua
-- 计算器类，用于构建和解析表达式
require("TerminalExpression")
require("NonterminalExpression")

Calculator = {}

function Calculator.calculate(expression)
    local endExpre = Calculator.createExpressionTree(expression, 1, #expression)
    return endExpre:interpret()
end

function Calculator.findSplit(expression, start, ended)
    local tag = -1
    if expression:sub(start, start) == '(' and expression:sub(ended, ended) == ')' then
        start = start + 1
        ended = ended - 1
    end
    local isInBracket = 0
    local moreGrade = 0
    for i = start, ended do
        if expression:sub(i, i) == '(' then
            isInBracket = isInBracket + 1
        elseif expression:sub(i, i) == ')' then
            isInBracket = isInBracket - 1
        end
        if (expression:sub(i, i) == '+' or expression:sub(i, i) == '-') and isInBracket == 0 then
            moreGrade = 1
            tag = i
        elseif (expression:sub(i, i) == '*' or expression:sub(i, i) == '/') and moreGrade == 0 and isInBracket == 0 then
            tag = i
        end
    end
    return tag
end

function Calculator.createExpressionTree(expression, start, ended)
    local ptree
    local substr = expression:sub(start, ended)
    local num = tonumber(substr)
    if num then
        ptree = NumberExpr:new(num)
    else
        local tag = Calculator.findSplit(expression, start, ended)
        if tag < 0 then
            -- 处理括号内的数字
            local s = string.gsub(string.gsub(substr, '%(', ''), '%)', '')
            num = tonumber(s)
            ptree = NumberExpr:new(num)
        else
            if expression:sub(tag, tag) == '+' then
                ptree = AdditionExpr:new(
                    Calculator.createExpressionTree(expression, start, tag - 1),
                    Calculator.createExpressionTree(expression, tag + 1, ended)
                )
            elseif expression:sub(tag, tag) == '-' then
                ptree = SubtractionExpr:new(
                    Calculator.createExpressionTree(expression, start, tag - 1),
                    Calculator.createExpressionTree(expression, tag + 1, ended)
                )
            elseif expression:sub(tag, tag) == '*' then
                ptree = MultiplicationExpr:new(
                    Calculator.createExpressionTree(expression, start, tag - 1),
                    Calculator.createExpressionTree(expression, tag + 1, ended)
                )
            elseif expression:sub(tag, tag) == '/' then
                ptree = DivisionExpr:new(
                    Calculator.createExpressionTree(expression, start, tag - 1),
                    Calculator.createExpressionTree(expression, tag + 1, ended)
                )
            end
        end
    end
    return ptree
end